app/services/agency_identity_linker.rb

```diff
@@ -0,0 +1,57 @@
+class AgencyIdentityLinker
+  def initialize(sp_identity)
+    @sp_identity = sp_identity
+    @agency_id = nil
+  end
+
+  def link_identity
+    ai = find_or_create_agency_identity
+    return ai if ai&.agency_enabled?
+    AgencyIdentity.new(user_id: @sp_identity.user_id, uuid: @sp_identity.uuid)
+  end
+
+  def self.sp_identity_from_uuid_and_sp(uuid, service_provider)
+    ai = AgencyIdentity.where(uuid: uuid).first
+    criteria = if ai&.agency_enabled?
+                 { user_id: ai.user_id, service_provider: service_provider }
+               else
+                 { uuid: uuid, service_provider: service_provider }
+               end
+    Identity.where(criteria).first
+  end
+
+  def self.sp_identity_from_uuid(uuid)
+    ai = AgencyIdentity.where(uuid: uuid).first
+    return Identity.where(uuid: uuid).first if ai.nil?
+    service_provider = ServiceProvider.where(agency_id: ai.agency_id).first
+    return unless service_provider
+    sp_identity_from_uuid_and_sp(ai.uuid, service_provider.issuer)
+  end
+
+  private
+
+  def find_or_create_agency_identity
+    ai = agency_identity
+    return ai if ai
+    create_agency_identity_for_sp
+  end
+
+  def create_agency_identity_for_sp
+    return unless agency_id
+    AgencyIdentity.create(agency_id: agency_id,
+                          user_id: @sp_identity.user_id,
+                          uuid: @sp_identity.uuid)
+  end
+
+  def agency_identity
+    ai = AgencyIdentity.where(uuid: @sp_identity.uuid).first
+    return ai if ai
+    sp = ServiceProvider.where(issuer: @sp_identity.service_provider).first
+    return unless agency_id(sp)
+    AgencyIdentity.where(agency_id: agency_id, user_id: @sp_identity.user_id).first
```

Comment:
How many agency identities can an agency have? If it's more than one, does the order matter here?

Reply:
There is only one agency_identity per user per agency.  The uuid is a unique key and so is the [agency_id,user_id]